### PR TITLE
[DeepSeek-V4.1] Bound dense prefill indexer memory

### PR DESCRIPTION
**Pull request by harmya**
## Motivation

The DeepSeek-V4.1 dense prefill indexer can OOM even when the KV cache has room. The HBM spike comes from temporary scores and candidate masks.

The inputs are FP4, but DeepGEMM returns FP32 scores with shape `[query tokens, visible context entries]`. At the candidate-source layer, those scores stay alive while we build Boolean masks in pieces and concatenate them. The pieces and the combined mask coexist during `torch.cat`.

```text
T = query tokens in the prefill chunk
N = visible context entries

FP32 scores          4 * T * N bytes
Boolean mask pieces  1 * T * N bytes
Combined mask        1 * T * N bytes
                     ---------------
Total                6 * T * N bytes

For a 16K chunk at 1M context:

6 * 16,384 * 1,048,576 / 2^30 = 96 GiB per GPU
```

This accounts for the main temporary allocations and not all model memory. These buffers are per rank, so TP4 does not mean we divide that 96 GiB by four.

## Modifications

Compute scores and token top-k in row tiles with a 2 GiB score budget. Each row still sees its full context. We keep batched GEMMs and top-k across request boundaries.

Keep the selected candidate block IDs between layers instead of a full Boolean mask. Build a mask only for the tile being processed.

```text
For 16K query rows and 2,048 candidate blocks per row:

Old mask at 1M context: 16,384 * 1,048,576 * 1 byte  = 16 GiB
New int32 block IDs:   16,384 *     2,048 * 4 bytes = 128 MiB
```

Candidate rows are sliced when entering the decoder replay tail. The compact format also works with replay disabled. Block IDs represent the selected mask losslessly. FP4 inputs and FP32 scores are unchanged. No flag is needed to enable this.

## Accuracy Tests

19 tests passed: 4 new CPU tests, 4 new SM100 GPU tests, and 11 existing nonpaged-indexer tests. Coverage includes ragged batches, empty requests, partial blocks, compression ratios 1 and 2, candidate consumers, replay tails, and peak memory. For tied scores, tests compare scores and valid selections rather than requiring one particular tie order.

## Speed Tests and Profiling

Indexer-only benchmark of the candidate-source path on one B200 with DeepGEMM 0.2.0 and PyTorch 2.13.0+cu130. Three warmups, then 15 measurements per version in alternating order. No allocator flush between calls.

| Query rows / visible entries | Baseline peak | Patched peak | Baseline median | Patched median |
| --- | ---: | ---: | ---: | ---: |
| 32 requests × 512 / 4K each | 0.34 GiB | 0.32 GiB | 5.81 ms | 5.66 ms |
| 4K / 128K | 3.38 GiB | 2.60 GiB | 12.08 ms | 10.37 ms |
| 16K / 128K | 12.00 GiB | 2.72 GiB | 47.21 ms | 40.50 ms |
| 16K / 512K | 48.00 GiB | 2.72 GiB | 186.48 ms | 165.58 ms |
| 16K / 1M | 96.01 GiB | 2.72 GiB | 372.14 ms | 330.18 ms |

Peak is extra PyTorch-allocated memory for scores, candidate selection, and token top-k. It excludes input tensors. The 2 GiB budget applies to scores only. Partial blocks can need a padded copy, so 2.72 GiB is not a bound for every shape.

Full-model cold-prefill results on 4xB200 with 16K chunks and `mem-fraction-static=0.70`:

| Prompt tokens | Baseline | Patched |
| ---: | ---: | ---: |
| 131,072 | 3.17 s | 3.44 s |
| 262,144 | 5.80 s | 6.10 s |
| 393,216 | 9.49 s | 9.30 s |
| 524,288 | OOM | 13.64 s |
| 786,432 | Not run | 24.56 s |
| 1,048,448 | Not run | 37.89 s |

Completed cold requests had zero cached tokens and generated 64 tokens. Times are end to end, with one run per length.

The baseline failed at `torch.cat(masks)` in `_publish_or_consume_candidates`. GPU 0 requested 7.25 GiB with 7.15 GiB free. The patched run completed every tested length and its cached repeat without OOM.

## Checklist

- [x] Full-repository pre-commit passed.
- [x] Registered CPU and GPU tests added and run.
- [x] Follow SGLang code style.
- [x] Reproduce the baseline OOM and run matching patched cold prefills.
- [ ] Finish serving-performance analysis and CI checks.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #35530026315](https://github.com/sgl-project/sglang/actions/runs/35530026315)
Latest PR Test (Extra): :white_check_mark: [Run #35530026049](https://github.com/sgl-project/sglang/actions/runs/35530026049)
Latest PR Test (AMD ROCm 10): :x: [Run #35530026146](https://github.com/sgl-project/sglang/actions/runs/35530026146)

### CODE REVIEW

**Comment by Oasis-Git** [comment]
Hi @harmya we have merged the pr https://github.com/sgl-project/sglang/pull/39095 which is highly related to your implementation. Can you make a rebase based on it then we can run CI? Thanks.

**Comment by harmya** [comment]
@Oasis-Git Rebased, thank you!

**Comment by Oasis-Git** [comment]
Hi Could you also reuse mqa_logits_rows_per_chunk() from #39095 in dense_prefill_indexer.py instead of duplicating the row-budget calculation? In high level #39095 should be a more general way to apply the chunk implementation.

**Comment by harmya** [comment]
Updated the PR @Oasis-Git! Happy to address more if needed :)

File: python/sglang/srt/layers/attention/dsv4/dense_prefill_indexer.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    mask_topk_scores,
+    select_candidate_block_ids,
+)
+from sglang.srt.layers.attention.mqa_logits_utils import (
+    mqa_logits_row_bytes,
+    mqa_logits_rows_per_chunk,
+)
+from sglang.srt.utils.common import ceil_align
+
+_SCORE_BUDGET_BYTES = 2 << 30
```
Comment (Oasis-Git) [inline python/sglang/srt/layers/attention/dsv4/dense_prefill_indexer.py:19]: It still uses a fixed 2 GiB budget. Would it make sense to use #39095’s free-memory-aware budget here as well? I think we do not need to change it in this pr since it should be merged asap but could u add a todo comment here?

**Review by Oasis-Git — APPROVED** [review]
Overall good to me. Maybe we should left one more todo comment. Let's run ci

File: python/sglang/srt/layers/attention/dsv4/dense_prefill_indexer.py (hunk 1 above)
Comment (harmya) [reply python/sglang/srt/layers/attention/dsv4/dense_prefill_indexer.py:19]: Yes, that makes sense. I have added that comment, I will keep an eye out for the follow up

**Comment by harmya** [comment]
@Oasis-Git Don't think I have perms to run CI. Would appreciate if you can run it 🙏🏻

**Comment by Oasis-Git** [comment]
Hi @harmya it is running now

**Comment by Oasis-Git** [comment]
Base pass: https://github.com/sgl-project/sglang/actions/runs/35530026315

**Comment by Oasis-Git** [comment]
Extra pass: https://github.com/sgl-project/sglang/actions/runs/35530026049
